universal-application-tool-0.0.1/app/repository/VersionRepository.java

```diff
@@ -0,0 +1,328 @@
+package repository;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Streams;
+import io.ebean.Ebean;
+import io.ebean.EbeanServer;
+import io.ebean.Transaction;
+import java.util.Optional;
+import java.util.stream.Stream;
+import javax.inject.Inject;
+import models.LifecycleStage;
+import models.Program;
+import models.Question;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import play.db.ebean.EbeanConfig;
+import services.program.BlockDefinition;
+import services.program.ProgramDefinition;
+import services.program.ProgramQuestionDefinition;
+
+public class VersionRepository {
+
+  private final EbeanServer ebeanServer;
+  private final ProgramRepository programRepository;
+  private final Logger LOG = LoggerFactory.getLogger(VersionRepository.class);
+
+  @Inject
+  public VersionRepository(EbeanConfig ebeanConfig, ProgramRepository programRepository) {
+    this.ebeanServer = Ebean.getServer(checkNotNull(ebeanConfig).defaultServer());
+    this.programRepository = checkNotNull(programRepository);
+  }
+
+  public void publishNewSynchronizedVersion() {
+    try {
+      Transaction transaction = ebeanServer.beginTransaction();
+      // For each unedited active program which does not have a draft program of the
+      // same name, set the version and go.
+      long version = getNextVersion();
+      for (Program uneditedActive :
+          ebeanServer
+              .find(Program.class)
+              .alias("active")
+              .where()
+              .eq("lifecycle_stage", LifecycleStage.ACTIVE)
+              .notExists(
+                  ebeanServer
+                      .find(Program.class)
+                      .where()
+                      .raw("active.name = name")
+                      .eq("lifecycle_stage", LifecycleStage.DRAFT)
+                      .query())
+              .findList()) {
+        uneditedActive.setLifecycleStage(LifecycleStage.OBSOLETE);
+        uneditedActive.save();
+        Program newVersion = new Program(uneditedActive.getProgramDefinition());
```

Comment:
Suggestion (non-blocking): L41 has a variable 'version' that's bound to a `long`, whereas `newVersion` here is bound to a program. Recommend renaming `newVersion` to `newProgramVersion`.

Reply:
Alternatively, rename `version` to `versionNumber`

Reply:
Went with `newActiveProgram`.

Reply:
Done.